### PR TITLE
Fix error handling in DTM retry logic and a bug in dispatcher's error handling

### PR DESCRIPTION
**Pull request by asimrp**
Please see individual commit messages for details.

### CODE REVIEW

File: src/backend/cdb/cdbtm.c [hunk 1]
```diff
@@ -791,7 +791,7 @@ doNotifyingCommitPrepared(void)
 	{
 		succeeded = doDispatchDtxProtocolCommand(DTX_PROTOCOL_COMMAND_COMMIT_PREPARED, /* flags */ 0,
 												 currentGxact->gid, currentGxact->gxid,
-												 &badGangs, /* raiseError */ false,
+												 &badGangs, /* raiseError */ true,
```
Comment (ashwinstar) [inline src/backend/cdb/cdbtm.c]: do we need to do that same for abort prepared as well ?

File: src/backend/cdb/dispatcher/cdbdispatchresult.c [hunk 2]
```diff
@@ -633,7 +633,10 @@ cdbdisp_get_PQerror(PGresult *pgresult)
 	if (fld)
 		errcontext("%s", fld);
 
-	return errfinish_and_return(0);
+	MemoryContext oldcontext = MemoryContextSwitchTo(TopTransactionContext);
```
Comment (ashwinstar) [inline src/backend/cdb/dispatcher/cdbdispatchresult.c]: I am not familiar with the usage of this interface, are we sure the returned pointer is not used like in abort case after the `TopTransactionContext` is gone.

File: src/test/regress/sql/dtm_retry.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
+-- Check if retry logic handles errors correctly.  The retry logic had
+-- a bug where error state wasn't cleaned up correctly during retries,
+-- leading to PANIC due to ERRORDATA_STACK_SIZE exeeded.
+set dtx_phase2_retry_count = 10;
```
Comment (ashwinstar) [inline src/test/regress/sql/dtm_retry.sql]: Can we make retry count to 11 and avoid panic, seems would avoid the need for `\c regression` which I feel maybe flaky depending on how long master takes to recover.

File: src/backend/cdb/cdbtm.c [hunk 4]
```diff
@@ -801,6 +801,7 @@ doNotifyingCommitPrepared(void)
 		 */
 		InterruptHoldoffCount = savedInterruptHoldoffCount;
 		succeeded = false;
+		FlushErrorState();
```
Comment (ashwinstar) [inline src/backend/cdb/cdbtm.c]: Check if same is needed for abort.

**Comment by ashwinstar** [comment]
Note: This issue exists in 5X_STABLE as well and hence needs to be backported once makes in master.

File: src/backend/cdb/dispatcher/cdbdispatchresult.c (hunk 2 above)
Comment (asimrp) [reply src/backend/cdb/dispatcher/cdbdispatchresult.c]: The `errfinish_and_return()` API is used only by dispatcher.  Of those usages, only two of them are actually using the returned error object.  And they are `doDispatchDtxProtocolCommand()` and `cdbCopyEndAndFetchRejectNum()`.  Both the calls happen within a transaction today.  I've added an assert and a comment to guide future uses of the API.

**Comment by asimrp** [comment]
Hmm, PR pipeline failed due to an unrelated issue.  Segment 1 could not come out of recovery after one of the isolation2 tests injected PANIC fault.  All subsequent tests failed.  From the container:
```
58473027e76d:/tmp/build/6d3dc0ce/gpdb_src/gpAux/gpdemo/datadirs$ PGOPTIONS='-c gp_session_role=utility' psql -d postgres -p 25433 
psql: FATAL:  the database system is in recovery mode
DETAIL:  last replayed record at 1/36B7E8
gpadmin@bf98b7a0-48fc-48b8-5ac5-58473027e76d:/tmp/build/6d3dc0ce/gpdb_src/gpAux/gpdemo/datadirs$ ps -ef | grep 25433
gpadmin    79215       0  0 02:21 ?        00:00:05 /usr/local/gpdb/bin/postgres -D /tmp/build/6d3dc0ce/gpdb_src/gpAux/gpdemo/datadirs/dbfast2/demoDataDir1 -p 25433 --gp_dbid=3 --gp_num_contents_in_cluster=3 --silent-mode=true --gp_contentid=1
gpadmin    79223   79215  0 02:21 ?        00:00:00 postgres: 25433, logger process   
gpadmin    80489   79215  0 02:23 ?        00:00:00 postgres: 25433, wal writer process   
gpadmin    98047   95065  0 03:07 pts/0    00:00:00 grep 25433
gpadmin@bf98b7a0-48fc-48b8-5ac5-58473027e76d:/tmp/build/6d3dc0ce/gpdb_src/gpAux/gpdemo/datadirs$
```
Unfortunately, this time, sending SIGABRT to the seemingly stuck wal writer process did not generate a core.  Also, no logs were generated in seg1's pg_log directory.  Does that mean logger process was also stuck?

We will never know for sure...

**Pull request merged**